### DISCUSSION TITLE
Output of custom Activity sent to HTTP Response asynchronously

### DISCUSSION THREAD

**Discussion by fulganta** [elsa-workflows/elsa-core#1579] [General]
Hi,
The goal of my workflow is to perform an asynchronous activity. 
We receive input from a http post which includes a callback url. We reply an correlation id of the current treatment, continue and post the result on the callback url provided by the user.

Here is the following activity I have:

1- HttpEndpoint - POST
![image](https://user-images.githubusercontent.com/45927310/134896765-67bfe94d-9492-4940-8aa8-ae55eb6f2d39.png)

2- SetVariable that store the post into a variable named quoteRequestInput. (Your example show to use "input.Body", but it can work just with "input" I think)
![image](https://user-images.githubusercontent.com/45927310/134896827-b6e558b5-e305-4a96-a63c-5c101e06dd29.png)

3- Custom Activity named GetCorrelationId that generate a unique ID and send it back to the caller that did the POST.
I tried to name the activity to be able to use it's output in the next activity as I have issue to catch it just writing "input". I seem to get the original post 1 activity input.
![image](https://user-images.githubusercontent.com/45927310/134898359-252c05a7-f4f6-44b0-87dc-20288b284eab.png)

The code include something simple here:
```
[ActivityOutput(Hint = "Correaltion ID", DefaultWorkflowStorageProvider = TransientWorkflowStorageProvider.ProviderName,
DisableWorkflowProviderSelection = true)]
public string CorrelationIdOuput { get; set; } = default!;

protected override async ValueTask<IActivityExecutionResult> OnExecuteAsync(ActivityExecutionContext context)
		{
			CorrelationIdOuput = "1234abcd";
			return Done();
		}
```

4- WriteHttpResponse that send back the Correl ID "1234abcd". 
I tried both JV and Liquid, but both seem to fail.
![image](https://user-images.githubusercontent.com/45927310/134898297-10c57c02-19fb-4481-ac70-c994776fb196.png)

5- Last activity that take the httpPost input and do something and post it to a callback URL
I seem to be able to get the correct data from quoteRequestInput, so all good here.
![image](https://user-images.githubusercontent.com/45927310/134898717-af8c17c2-b813-4c48-8a6d-38480b28c5e2.png)

So 2 questions:

1- How can I  catch the output of my GetCorrelationId activity named CorrelationIdOuput (step 3)  in my WriteHttpResponse (step 4) ? I don't seem to be able to do it even if follow your tutorial.

2- It seems that the output WriteHttpResponse  is sent after execution of activity 5, so it's not really asynchronous here... which is the goal. How can I reply to the original post before the all workflow is executed?

**Comment by sfmskywalker** [comment]
Hi @cfrelet ,

1. All looks good, so I'm surprised that the output doesn't render the expected correlation ID. By the way, if you rename your `CorrelationIdOuput` property to `Output`, it will become available via the `input` JS and `Input` Liquid variables. However, even with your current code you should definitely be able to do `Activities.GenCorrActivity.CorrelationIdOuput` and get the generated value printed to the response, so something else must be going on.
2. The Write HTTP Response doesn't terminate the current request, it will simply do `Response.WriteAsync` internally and the workflow engine will continue executing subsequent activities until it either encounters a blocking activity or reaches the end of the workflow. This will be configurable with #1144. Until that's done, one thing you can do to continue the workflow asynchronously is to add a blocking activity right after the Write HTTP Response activity. For example, you could add a Timer activity (make sure to enter a non-zero timeout, like 1 second or less, but greater than 0). This will cause the workflow to get suspended and then resumed immediately from a background process.

Regarding the input issue, feel free to send me a (simplified) project that demonstrates the issue.

**Reply by fulganta** [reply]
HI @sfmskywalker ,
Thanks a lot for your quick feedback as usual.
1- Replacing `CorrelationIdOuput ` by `Output` fix the issue indeed. I didn't know we have to stick out that name to use it in JS and Liquid.
2- This trick work indeed with a Timer thank you. I'm looking forward to be able to do it natively with HTTP answer.

**Reply by sfmskywalker** [reply]
>  Replacing CorrelationIdOuput by Output fix the issue indeed. I didn't know we have to stick out that name to use it in JS and Liquid.

Technically you can use any name you want, including `CorrelationIdOuput`. But any output property that is named `Output` will be automatically passed as input into the next activity, made available via the global `input` JS variable or `Input` Liquid variable. 

But if you name it `CorrelationIdOuput` for example, you can still access it from other activities, just using a different syntax, e.g.: `activities.GenCorrActivity.CorrelationIdOuput()`. 

It's strange that `Input` works for you, while `Activities.GenCorrActivity.CorrelationIdOuput` does not.

If you want to send me a repro I'd be happy to investigate.

**Reply by fulganta** [reply]
Maybe i messed up somewhere….
One more question, can I take multiple input in an activity? In my case I want to take the correlationID generated + the quoterequestInput I get from the original user Post. I guess it’s possible, maybe just need separate by coma the parameters?

**Reply by fulganta** [reply]
Ok I found a way to have 2 input in a activity.
I just have to double the activity tag:
```
[ActivityInput(Label = "Quote Request",
			Hint = "Quote Request details",
			SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid })]
public Dictionary<string, string> quoteRequestInput { get; set; } = default!;

[ActivityInput(Label = "Correlation ID",
			Hint = "input correl id generated",
			SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid })]
public string corridVal { get; set; } = default!;
```
The input comes from a first post I get that I save in SetVariable activity as `quoteRequestInput`
![image](https://user-images.githubusercontent.com/45927310/135421217-c2229ca5-8058-4c39-99d8-61e37b6d92a9.png)
The second input comes from the output from another activity that I catch also with a SetVariable activity as `corridVal`
![image](https://user-images.githubusercontent.com/45927310/135421513-60446fae-2c78-44a2-9176-34fabccdcd3f.png)

Note that at the input activity level I add `quoteRequestInput` as a Javascript as it's a dictionary and I didn't succeed to make it with Liquid.
But for the other input `corridVal` I use Liquid expression.
![image](https://user-images.githubusercontent.com/45927310/135421818-c2d5e98b-0b2d-473f-8bdf-f7145f5b5f3f.png)
Let me know if there was a simpler way to do it.

**Reply by sfmskywalker** [reply]
I can't think of a better way. An alternative solution might be the use of the `RunJavaScript` activity to set both variables from a single script, reducing the two `SetVariable` activities to a single `RunJavaScript` that sets two variables (using the `setVariable` function).
